File: implementations/lua/rcl/converters.lua
local function named_base(name) return name == "region" and "regions" or name end

local function esc(s)
  return s:gsub("\\", "\\\\"):gsub('"', '\\"'):gsub("\n", "\\n"):gsub("\t", "\\t")
end

local function node_value(node)
  if node.kind == "array" then local out = {}; for i, e in ipairs(node.elements) do out[i] = node_value(e) end; return out end
  return node.value
end

local function insert_path(dst, key, value)
  local parts = {}
  for p in key:gmatch("[^.]+") do parts[#parts + 1] = p end
  local cur = dst
  for i = 1, #parts - 1 do
    local p = parts[i]
    if cur[p] == nil then cur[p] = {} end
    if type(cur[p]) ~= "table" then error("key conflict at '" .. p .. "'") end
    cur = cur[p]
  end
  if cur[parts[#parts]] ~= nil then error("duplicate key '" .. key .. "'") end
  cur[parts[#parts]] = value
end

local function block_object(block)
  local out = {}
  for _, k in ipairs(block.property_order) do insert_path(out, k, node_value(block.properties[k])) end
  for _, b in ipairs(block.blocks) do
    local child = block_object(b)
    if type(out[b.name]) == "table" then for k, v in pairs(child) do out[b.name][k] = v end else out[b.name] = child end
  end
  for _, b in ipairs(block.named_blocks) do
    local base = named_base(b.name)
    if type(out[base]) ~= "table" then out[base] = {} end
    out[base][b.argument] = block_object(b)
  end
  return out
end

local function scalar(v)
  if type(v) == "string" then return '"' .. esc(v) .. '"' end
  if type(v) == "boolean" then return v and "true" or "false" end
  if type(v) == "number" then return tostring(v) end
  if type(v) == "table" and #v > 0 then local out = {}; for _, e in ipairs(v) do out[#out + 1] = scalar(e) end; return "[" .. table.concat(out, ", ") .. "]" end
  if type(v) == "table" then return "{}" end
  return "null"
end

local function sorted_keys(t)
  local keys = {}
  for k, _ in pairs(t) do keys[#keys + 1] = k end
  table.sort(keys)
  return keys
end

local function emit_yaml(v, indent)
  local pad = string.rep("  ", indent)
  if type(v) == "table" and #v > 0 then
    local lines = {}
    for _, e in ipairs(v) do
      if type(e) == "table" then lines[#lines + 1] = pad .. "-"; lines[#lines + 1] = emit_yaml(e, indent + 1)
      else lines[#lines + 1] = pad .. "- " .. scalar(e) end
    end
    return table.concat(lines, "\n")
  elseif type(v) == "table" then
    local lines = {}
    for _, k in ipairs(sorted_keys(v)) do
      local item = v[k]
      if type(item) == "table" then lines[#lines + 1] = pad .. k .. ":"; lines[#lines + 1] = emit_yaml(item, indent + 1)
      else lines[#lines + 1] = pad .. k .. ": " .. scalar(item) end
    end
    return table.concat(lines, "\n")
  end
  return pad .. scalar(v)
end

local function emit_toml(v)
  local out = {}
  local function walk(obj, prefix)
    for _, k in ipairs(sorted_keys(obj)) do if type(obj[k]) ~= "table" or #obj[k] > 0 then out[#out + 1] = k .. " = " .. scalar(obj[k]) end end
    for _, k in ipairs(sorted_keys(obj)) do
      local item = obj[k]
      if type(item) == "table" and #item == 0 then
        local sec = prefix and (prefix .. "." .. k) or k
        if #out > 0 then out[#out + 1] = "" end
        out[#out + 1] = "[" .. sec .. "]"
        walk(item, sec)
      end
    end
  end
  walk(v, nil)
  return table.concat(out, "\n")
end

local function emit_hcl(v, indent)
  local pad, lines = string.rep("  ", indent), {}
  for _, k in ipairs(sorted_keys(v)) do
    local item = v[k]
    if type(item) == "table" and #item == 0 then lines[#lines + 1] = pad .. k .. " {"; lines[#lines + 1] = emit_hcl(item, indent + 1); lines[#lines + 1] = pad .. "}"
    else lines[#lines + 1] = pad .. k .. " = " .. scalar(item) end
  end
  return table.concat(lines, "\n")
end

return {
  to_object = function(doc)
    local out = {}
    for _, b in ipairs(doc.blocks) do
      if b.argument then out[named_base(b.name)] = { [b.argument] = block_object(b) }
      else out[b.name] = block_object(b) end
    end
    return out
  end,
  to_yaml = function(obj) return emit_yaml(obj, 0) end,
  to_toml = function(obj) return emit_toml(obj) end,
  to_hcl = function(obj) return emit_hcl(obj, 0) end,
}

File: implementations/julia/src/rcl/Converters.jl
named_base(name::String) = endswith(name, "s") ? name : string(name, "s")
node_value(node) = node["kind"] == "array" ? map(node_value, node["elements"]) : node["kind"] == "block" ? block_object(node["block"]) : node["value"]

function insert_path!(dst::Dict{String,Any}, key::String, value)
  parts = split(key, ".")
  cur = dst
  for i in 1:length(parts)-1
    p = parts[i]
    if !haskey(cur, p); cur[p] = Dict{String,Any}() end
    cur[p] isa Dict{String,Any} || error("key conflict at '$p'")
    cur = cur[p]
  end
  haskey(cur, parts[end]) && error("duplicate key '$key'")
  cur[parts[end]] = value
end

function block_object(block)
  out = Dict{String,Any}()
  for key in block["property_order"]; insert_path!(out, key, node_value(block["properties"][key])) end
  for child in block["blocks"]
    c = block_object(child)
    if get(out, child["name"], nothing) isa Dict{String,Any}; merge!(out[child["name"]], c) else out[child["name"]] = c end
  end
  for child in block["named_blocks"]
    base = named_base(child["name"])
    if !(get(out, base, nothing) isa Dict{String,Any}); out[base] = Dict{String,Any}() end
    out[base][child["argument"]] = block_object(child)
  end
  out
end

function to_object_doc(doc)
  if haskey(doc, "root_value") && !isnothing(doc["root_value"])
    return Dict{String,Any}("root" => node_value(doc["root_value"]))
  end
  out = Dict{String,Any}()
  for block in doc["blocks"]
    if isnothing(block["argument"]); out[block["name"]] = block_object(block)
    else out[named_base(block["name"])] = Dict{String,Any}(block["argument"] => block_object(block)) end
  end
  out
end

esc_scalar(s::String) = "\"" * replace(replace(replace(replace(s, "\\"=>"\\\\"), "\""=>"\\\""), "\n"=>"\\n"), "\t"=>"\\t") * "\""
scalar(v) = v isa String ? esc_scalar(v) : v isa Bool ? (v ? "true" : "false") : v isa Number ? string(v) : v isa Vector ? "[" * join(map(scalar, v), ", ") * "]" : v isa Dict{String,Any} ? "{}" : "{}"

function emit_yaml(v, indent=0)
  pad = repeat("  ", indent)
  if v isa Vector
    return join(map(x -> x isa Dict{String,Any} || x isa Vector ? "$(pad)-\n$(emit_yaml(x, indent + 1))" : "$(pad)- $(scalar(x))", v), "\n")
  elseif v isa Dict{String,Any}
    keys_sorted = sort(collect(keys(v)))
    return join(map(k -> begin item = v[k]; item isa Dict{String,Any} || item isa Vector ? "$(pad)$(k):\n$(emit_yaml(item, indent + 1))" : "$(pad)$(k): $(scalar(item))" end, keys_sorted), "\n")
  end
  pad * scalar(v)
end

function emit_toml(v::Dict{String,Any})
  out = String[]
  function walk(obj::Dict{String,Any}, prefix=nothing)
    for k in sort(collect(keys(obj)))
      item = obj[k]
      (item isa Dict{String,Any}) && continue
      push!(out, "$k = $(scalar(item))")
    end
    for k in sort(collect(keys(obj)))
      item = obj[k]
      (item isa Dict{String,Any}) || continue
      sec = isnothing(prefix) ? k : "$(prefix).$(k)"
      !isempty(out) && push!(out, "")
      push!(out, "[$sec]")
      walk(item, sec)
    end
  end
  walk(v)
  join(out, "\n")
end

function emit_hcl(v::Dict{String,Any}, indent=0)
  pad, lines = repeat("  ", indent), String[]
  for k in sort(collect(keys(v)))
    item = v[k]
    if item isa Dict{String,Any}
      push!(lines, "$(pad)$(k) {")
      push!(lines, emit_hcl(item, indent + 1))
      push!(lines, "$(pad)}")
    else
      push!(lines, "$(pad)$(k) = $(scalar(item))")
    end
  end
  join(lines, "\n")
end
